File: 0091-decode-ways/0091-decode-ways.cpp
// Recursive memoization solution

class Solution {
private:
    vector<int> dp;
public:
    int ways(int pos, string &s) {
        if (pos >= s.size()) {
            return 1;
        }
        if (dp[pos] != -1) {
            return dp[pos];
        }
        int take1 = 0, take2 = 0;
        if (s[pos] != '0') {
            take1 = ways(pos + 1, s);
            if (pos + 1 < s.size() ) {
                int number = stoi(s.substr(pos, 2));
                if (number >= 10 && number <= 26) {
                    take2 = ways(pos + 2, s);
                }
            }
        }
        return dp[pos] = take1 + take2;
    }
    int numDecodings(string s) {
        dp.clear();
        dp.resize(s.size(), -1);
        return ways(0, s);
    }
};

// Time: O(N)
// Space: O(N)

// // Iterative solution

// class Solution {
// public:
//     int numDecodings(string s) {
//         int len = s.size();
//         vector<int> dp(len + 1);
//         dp[0] = 1;
//         dp[1] = s[0] == '0' ?  0 : 1;
//         for (int i = 2; i <= len; i++) {
//             if (s[i - 1] != '0') {
//                 dp[i] += dp[i - 1];
//             }
//             int number = stoi(s.substr(i - 2, 2));
//             if (number >= 10 && number <= 26) {
//                 dp[i] += dp[i - 2];
//             }
//         }
//         return dp[len];
//     }
// };

// // Time: O(N)
// // Space: O(N)

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        # dp = [-1] * len(s)
        # def ways(pos):
        #     if pos >= len(s):
        #         return 1
        #     if dp[pos] != -1:
        #         return dp[pos]
        #     take1, take2 = 0, 0
        #     if s[pos] != '0':
        #         take1 = ways(pos + 1)
        #         if pos + 1 < len(s):
        #             num = int(s[pos:pos + 2])
        #             if num <= 26:
        #                 take2 = ways(pos + 2)
        #     dp[pos] = take1 + take2
        #     return dp[pos]
        # return ways(0)
        n = len(s)
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 0 if s[0] == '0' else 1
        for i in range(2, n + 1):
            if s[i - 1] != '0':
                dp[i] += dp[i - 1]
            num = int(s[i - 2:i])
            if 10 <= num <= 26:
                dp[i] += dp[i - 2]
        return dp[n]
                    